File: parser.cpp
// system includes
#include <map>
#include <memory>

// includes
#include "parser.h"
#include "basic_objects/parser.h"
#include "basic_parser/parser.h"
#include "basic_parser/malformed_request.h"
#include "validator.h"
#include "request_type_parser.h"

namespace user_management_protocol
{

namespace parser
{

using ::basic_parser::parser::get_value_or_throw;
using ::basic_parser::parser::get_value_or_throw_t;

// enums

void get_value_or_throw( gender_e * res, const std::string & key, const generic_request::Request & r )
{
    uint32_t res_i;

    get_value_or_throw( & res_i, key, r );

    * res = static_cast<gender_e>( res_i );
}

// objects

void get_value_or_throw( UserInfo * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->gender, prefix + ".GENDER", r );
    get_value_or_throw( & res->last_name, prefix + ".LAST_NAME", r );
    get_value_or_throw( & res->first_name, prefix + ".FIRST_NAME", r );
    get_value_or_throw( & res->company_name, prefix + ".COMPANY_NAME", r );
    get_value_or_throw( & res->email, prefix + ".EMAIL", r );
    get_value_or_throw( & res->email_2, prefix + ".EMAIL_2", r );
    get_value_or_throw( & res->phone, prefix + ".PHONE", r );
    get_value_or_throw( & res->phone_2, prefix + ".PHONE_2", r );
    get_value_or_throw( & res->timezone, prefix + ".TIMEZONE", r );
}

// base messages

void get_value_or_throw( Request * res, const generic_request::Request & r )
{
}

void get_value_or_throw( BackwardMessage * res, const generic_request::Request & r )
{
}

// messages

void get_value_or_throw( SetUserInfoRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<Request*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
    get_value_or_throw( & res->user_info, "USER_INFO", r );
}

void get_value_or_throw( SetUserInfoResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

}

void get_value_or_throw( GetUserInfoRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<Request*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
}

void get_value_or_throw( GetUserInfoResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
    get_value_or_throw( & res->gender, "GENDER", r );
    get_value_or_throw( & res->last_name, "LAST_NAME", r );
    get_value_or_throw( & res->first_name, "FIRST_NAME", r );
    get_value_or_throw( & res->company_name, "COMPANY_NAME", r );
    get_value_or_throw( & res->email, "EMAIL", r );
    get_value_or_throw( & res->email_2, "EMAIL_2", r );
    get_value_or_throw( & res->phone, "PHONE", r );
    get_value_or_throw( & res->phone_2, "PHONE_2", r );
    get_value_or_throw( & res->timezone, "TIMEZONE", r );
}

// to object

Object * to_SetUserInfoRequest( const generic_request::Request & r )
{
    std::unique_ptr<SetUserInfoRequest> res( new SetUserInfoRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_SetUserInfoResponse( const generic_request::Request & r )
{
    std::unique_ptr<SetUserInfoResponse> res( new SetUserInfoResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetUserInfoRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetUserInfoRequest> res( new GetUserInfoRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetUserInfoResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetUserInfoResponse> res( new GetUserInfoResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

// to forward message

basic_parser::Object* to_forward_message( const generic_request::Request & r )
{
    auto type = detect_request_type( r );

    typedef request_type_e KeyType;

    typedef Object* (*PPMF)( const generic_request::Request & r );

#define HANDLER_MAP_ENTRY(_v)       { KeyType::_v,    & to_##_v }

    static const std::map<KeyType, PPMF> funcs =
    {
        HANDLER_MAP_ENTRY( SetUserInfoRequest ),
        HANDLER_MAP_ENTRY( SetUserInfoResponse ),
        HANDLER_MAP_ENTRY( GetUserInfoRequest ),
        HANDLER_MAP_ENTRY( GetUserInfoResponse ),
    };

#undef HANDLER_MAP_ENTRY

    auto it = funcs.find( type );

    if( it != funcs.end() )
        return it->second( r );

    return nullptr;
}

using basic_parser::MalformedRequest;

request_type_e  detect_request_type( const generic_request::Request & r )
{
    std::string cmd;

    if( r.get_value( "CMD", cmd ) == false )
        throw MalformedRequest( "CMD is not defined" );

    return RequestTypeParser::to_request_type( cmd );
}

} // namespace parser

} // namespace user_management_protocol

File: parser.php
<?php

namespace user_management_protocol;


// base include
require_once __DIR__.'/../generic_protocol/parser.php';
// includes
require_once __DIR__.'/../basic_objects/parser.php';
require_once __DIR__.'/../basic_parser/parser.php';

// enums

function parse__gender_e( & $csv_arr, & $offset )
{
    $res = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

// objects

function parse__UserInfo( & $csv_arr, & $offset )
{
    $res = new UserInfo;

    $res->gender = parse__gender_e( $csv_arr, $offset );
    $res->last_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->first_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->company_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->email = \basic_parser\parse__string( $csv_arr, $offset );
    $res->email_2 = \basic_parser\parse__string( $csv_arr, $offset );
    $res->phone = \basic_parser\parse__string( $csv_arr, $offset );
    $res->phone_2 = \basic_parser\parse__string( $csv_arr, $offset );
    $res->timezone = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

// base messages

function parse__Request( & $res, & $csv_arr, & $offset )
{
    // no base class

}

function parse__BackwardMessage( & $res, & $csv_arr, & $offset )
{
    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );


}

// messages

function parse__SetUserInfoRequest( & $csv_arr )
{
    $res = new SetUserInfoRequest;

    $offset = 1;

    // base class
    parse__Request( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->user_info = parse__UserInfo( $csv_arr, $offset );

    return $res;
}

function parse__SetUserInfoResponse( & $csv_arr )
{
    $res = new SetUserInfoResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );


    return $res;
}

function parse__GetUserInfoRequest( & $csv_arr )
{
    $res = new GetUserInfoRequest;

    $offset = 1;

    // base class
    parse__Request( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

function parse__GetUserInfoResponse( & $csv_arr )
{
    $res = new GetUserInfoResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->gender = parse__gender_e( $csv_arr, $offset );
    $res->last_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->first_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->company_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->email = \basic_parser\parse__string( $csv_arr, $offset );
    $res->email_2 = \basic_parser\parse__string( $csv_arr, $offset );
    $res->phone = \basic_parser\parse__string( $csv_arr, $offset );
    $res->phone_2 = \basic_parser\parse__string( $csv_arr, $offset );
    $res->timezone = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

// generic

class Parser extends \generic_protocol\Parser
{

protected static function parse_csv_array( $csv_arr )
{
    if( sizeof( $csv_arr ) < 1 )
        return NULL;

    $handler_map = array(
        // messages
        'user_management_protocol/SetUserInfoRequest'         => 'parse__SetUserInfoRequest',
        'user_management_protocol/SetUserInfoResponse'         => 'parse__SetUserInfoResponse',
        'user_management_protocol/GetUserInfoRequest'         => 'parse__GetUserInfoRequest',
        'user_management_protocol/GetUserInfoResponse'         => 'parse__GetUserInfoResponse',
    );

    $type = $csv_arr[0][0];

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\user_management_protocol\\' . $handler_map[ $type ];
        return $func( $csv_arr[0] );
    }

    return \generic_protocol\Parser::parse_csv_array( $csv_arr );
}

}

// namespace_end user_management_protocol


?>
